src/Npgsql/NpgsqlRecordDataReader.cs

```diff
@@ -0,0 +1,327 @@
+﻿using Npgsql.Internal;
+using Npgsql.Internal.TypeHandling;
+using Npgsql.PostgresTypes;
+using Npgsql.TypeMapping;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Npgsql
+{
+    internal sealed class NpgsqlRecordDataReader : DbDataReader
+    {
+        NpgsqlDataReader _outermostReader;
+        ulong _uniqueOutermostReaderRowId;
+        int _depth;
+        int _numRows;
+        int _nextRowIndex;
+        int _nextRowBufferPos;
+
+        ColumnInfo[]? _columns;
+
+        bool _calledNextResult;
+        bool _closed;
+
+        struct ColumnInfo
+        {
+            public uint Oid;
+            public int BufferPos;
+        }
+
+        NpgsqlReadBuffer Buffer => _outermostReader.Buffer;
+        ConnectorTypeMapper TypeMapper => _outermostReader.Connector.TypeMapper;
+
+        internal NpgsqlRecordDataReader(NpgsqlDataReader outermostReader, ulong uniqueOutermostReaderRowId, int depth)
+        {
+            _outermostReader = outermostReader;
+            _uniqueOutermostReaderRowId = uniqueOutermostReaderRowId;
+            _depth = depth;
+        }
+
+        internal void InitArray()
+        {
+            var dimensions = Buffer.ReadInt32();
+            var containsNulls = Buffer.ReadInt32() == 1;
+            Buffer.ReadUInt32(); // Element OID. Ignored.
+
+            if (containsNulls)
+                throw new InvalidOperationException("Record array contains null record");
+
+            if (dimensions == 0)
+                return;
+
+            if (dimensions != 1)
+                throw new InvalidOperationException($"Cannot read an array with 1 dimension from an array with {dimensions} dimension(s)");
```

Comment:
Maybe adapt the message bit more: 

```suggestion
                throw new InvalidOperationException("Cannot read a multidimensional array with an inner NpgsqlDataReader");
```

Reply:
Maybe "Cannot read a multidimensional array with an inner DbDataReader"?

Reply:
Sure.